p2p/dht/dht.go

```diff
@@ -185,57 +226,90 @@ func filterFindNodeServers(nodes []node.Node, queried map[string]struct{}, alpha
 	return newlist
 }
 
+type findNodeOpRes struct {
+	res    []node.Node
+	server node.Node
+}
+
 // findNodeOp a target node on one or more servers
 // returns closest nodes which are closers than closestNode to targetId
 // if node found it will be in top of results list
-func (d *KadDHT) findNodeOp(servers []node.Node, queried map[string]struct{}, id p2pcrypto.PublicKey,
-	closestNode node.Node) []node.Node {
+func (d *KadDHT) findNodeOp(servers []node.Node, queried map[string]bool, id p2pcrypto.PublicKey) []node.Node {
+
+	var out []node.Node
+	startTime := time.Now()
+
+	defer func() {
+		d.local.With().Debug("findNodeOp", log.Int("servers", len(servers)), log.Int("result_count", len(out)), log.Duration("time_elapsed", time.Since(startTime)))
+	}()
 
 	l := len(servers)
 
 	if l == 0 {
 		return []node.Node{}
 	}
 
-	// results channel
-	results := make(chan []node.Node)
+	results := make(chan *findNodeOpRes)
+
+	// todo : kademlia says: This recursion can begin before all of the previous RPCs have returned
+	// currently we wait for all.
 
 	// Issue a parallel FindNode op to all servers on the list
 	for i := 0; i < l; i++ {
-		server := servers[i]
-		queried[server.String()] = struct{}{}
-		idx := id
+
 		// find node protocol adds found nodes to the local routing table
 		// populates queried node's routing table with us and return.
-		go func(server node.Node, id p2pcrypto.PublicKey) {
-			fnd, err := d.fnp.FindNode(server, id)
+		go func(server node.Node, idx p2pcrypto.PublicKey) {
+			d.rt.Update(server) // must do this prior to message to make sure we have that node
+
+			var err error
+
+			defer func() {
+				if err != nil {
+					d.local.With().Debug("find_node_failed", log.String("server", server.String()), log.Err(err))
+					results <- &findNodeOpRes{nil, server}
+					d.rt.Remove(server) // remove node if it didn't work
+					return
+				}
+			}()
+
+			err = d.ping.Ping(server.PublicKey())
+			if err != nil {
+				return
+			}
+
+			fnd, err := d.fnp.FindNode(server, idx)
 			if err != nil {
-				//TODO: handle errors
 				return
 			}
-			results <- fnd
-		}(server, idx)
+
+			results <- &findNodeOpRes{fnd, server}
+		}(servers[i], id)
 	}
 
 	done := 0                          // To know when all operations finished
 	idSet := make(map[string]struct{}) // to remove duplicates
 
-	out := make([]node.Node, 0) // the end result we collect
-
 	timeout := time.NewTimer(LookupTimeout)
 Loop:
 	for {
 		select {
-		case res := <-results:
+		case qres := <-results:
+
+			// we mark active nodes
+			queried[qres.server.String()] = qres.res == nil
```

Comment:
queried should be true if result is not nil
`queried[qres.server.String()] = qres.res != nil`